### PR TITLE
Add ability to exclude sample results to be sent as logs based on response code regex

### PR DESCRIPTION
**Pull request by arustamov**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/README.md#contributing). 

### What does this PR do?

https://github.com/DataDog/jmeter-datadog-backend-listener/issues/47

### Description of the Change

Screenshot 2023-11-08 at 19 55 34
Screenshot 2023-11-07 at 17 08 30
Screenshot 2023-11-07 at 17 08 35
Screenshot 2023-11-07 at 17 08 43
Screenshot 2023-11-07 at 17 08 53
Screenshot 2023-11-07 at 17 09 07
Screenshot 2023-11-07 at 17 08 12

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

**Review by alopezz — COMMENTED** [review]
Thanks for your contribution, this looks helpful! However:

- Can we avoid all the formatting changes in unrelated code? They make the diff larger than necessary and harder to spot what has actually change.
- Can we add tests for the new feature? I see there's a new test but it only checks for configuration errors.
- I wonder if it would make sense to make this a bit more flexible and configurable w.r.t. which codes result in logs being sent. What do you think?

**Comment by arustamov** [comment]
Hi @alopezz , thanks for the review

1. Pushed reverting extra code formatting applied by the default IDEA config
2. The new test testSendOnlyErrorsAsLogsNotBoolean has been added similar to the other ones boolean-based like testSendResultsAsLogsNotBoolean and testIncludeSubresultsNotBoolean.
I don't see any sophisticated test coverage around this so I would really like please not overcomplicate for the specific feature being added now
3. As to the exact solution, I was considering going with regex pattern-based config similar to the sampler filter. The key difference here is that we have not only HTTP code responses but also non-HTTP ones. 
So going with boolean solution saves us from rather complicated pattern to put here as the default value. In addition to this possible non-HTTP responses might have values that are hard to know in advance to config to be able to capture as logs

**Comment by arustamov** [comment]
> Thanks for your contribution, this looks helpful! However:
> 
> * Can we avoid all the formatting changes in unrelated code? They make the diff larger than necessary and harder to spot what has actually change.
> * Can we add tests for the new feature? I see there's a new test but it only checks for configuration errors.
> * I wonder if it would make sense to make this a bit more flexible and configurable w.r.t. which codes result in logs being sent. What do you think?

**Comment by arustamov** [comment]
Adding the default behavior as an extra evidense
Screenshot 2023-11-08 at 14 26 08
Screenshot 2023-11-08 at 14 29 31
Screenshot 2023-11-08 at 14 30 03

**Comment by alopezz** [comment]
> 1. Pushed reverting extra code formatting applied by the default IDEA config

I still see many formatting changes that are unrelated to the PR's goal (the markdown files, the import order somewhere else...). I apologize for not having an automated formatting setup for this project at this point; any "general reformatting" change should be in a separate PR if we ever set that up, not a part of a feature PR.

> 2. The new test testSendOnlyErrorsAsLogsNotBoolean has been added similar to the other ones boolean-based like testSendResultsAsLogsNotBoolean and testIncludeSubresultsNotBoolean.
>    I don't see any sophisticated test coverage around this so I would really like please not overcomplicate for the specific feature being added now

We do have tests for similar cases. See, for example, https://github.com/DataDog/jmeter-datadog-backend-listener/blob/a18e76ec25e121f31207642c895e208a13a15a47/src/test/java/org/datadog/jmeter/plugins/DatadogBackendClientTest.java#L191-L197

Writing tests is part of our [contributing guidelines](https://github.com/DataDog/jmeter-datadog-backend-listener#pull-requests) and a good way to ensure the feature works as advertised without having to manually test it.

> 3. As to the exact solution, I was considering going with regex pattern-based config similar to the sampler filter. The key difference here is that we have not only HTTP code responses but also non-HTTP ones.
>    So going with boolean solution saves us from rather complicated pattern to put here as the default value. In addition to this possible non-HTTP responses might have values that are hard to know in advance to config to be able to capture as logs

That's the kind of idea that I had in mind, too. A way that this might work is if we have a regex pattern for exclusion, defaulting to an empty string (meaning the current default behavior of sending all results), and documenting the pattern for errors as an example. Wouldn't that work well enough? Otherwise we're setting on a definition of "error" that I'm not sure is general enough (though I'm not a JMeter user so I may miss some context).

File: README.md [hunk 1]
```diff
@@ -46,6 +46,7 @@ The plugin has the following configuration options:
 |logsBatchSize|false|500|Logs are submitted in batches of size `logsBatchSize` as soon as this size is reached.|
 |sendResultsAsLogs|false|false|By default only metrics are reported to Datadog. To report individual test results as log events, set this field to `true`.|
 |includeSubresults|false|false|A subresult is for instance when an individual HTTP request has to follow redirects. By default subresults are ignored.|
+|excludeLogsResponseCodeRegex|false|`""`| By default if send test results as log events enabled, then all logs are reported to Datadog. To exclude based on response code of sample result, set regex here. i.e. `[123][0-5][0-9]` to send only errors|
```
Comment (alopezz) [inline README.md:49]:
```suggestion
|excludeLogsResponseCodeRegex|false|`""`| Setting `sendResultsAsLogs` will submit all results as logs to Datadog by default. This option lets you exclude results whose response code matches a given regex. For example, you may set this option to `[123][0-5][0-9]` to only submit errors.|
```

**Review by alopezz — CHANGES_REQUESTED** [review]
Thanks, nice work! This is pretty much ready to go, just a suggestion to improve the option's description and we'll merge it.

**Pull request merged**